src/MemoryRouter.tsx

```diff
@@ -145,21 +146,36 @@ export class MemoryRouter extends BaseRouter {
     const shallow = options?.shallow || false;
     const pathname = removeTrailingSlash(urlObject.pathname || "");
     const query = urlObject.query || {};
-    const asPath = getRouteAsPath(baseUrlObject.pathname ?? "", baseQuery, urlObject.hash);
-
-    this.events.emit("routeChangeStart", asPath, { shallow });
+    const hash = urlObject.hash || "";
+    const asPath = getRouteAsPath(baseUrlObject.pathname ?? "", baseQuery, hash);
+
+    const isHashChange = this.hash !== hash;
+    const isQueryChange = stringifyQueryString(this.query) !== stringifyQueryString(query);
+    const isRouteChange = isQueryChange || this.pathname !== pathname;
+    const isOnlyHashChange = isHashChange && !isRouteChange;
```

Comment:
Thank you so much for the codesandbox (I forgot that they supported Next apps, because otherwise it's tedious to set up my own testing!).
Playing with it, I found there was a little bit of edge-case behavior, and perhaps we should try to capture that logic.
The edge-case is when I call `push('/foo#bar')` twice.  In Next, this triggers the hash-only change, even though the hash did NOT change.  So, to be as accurate as possible, I think the logic should be:

```suggestion
    const isQueryChange = stringifyQueryString(this.query) !== stringifyQueryString(query);
    const isRouteChange = isQueryChange || this.pathname !== pathname;
    const isOnlyHashChange = !isRouteChange && hash;
```

Reply:
Unfortunately there is one more edge case - removing the hash (e.g. `/foo#hash` -> `/foo`) will also trigger the hashChange:

```
/foo
  -> routeChange
/foo#baz
  -> hashChange
/foo#baz
  -> hashChange
/foo
  -> hashChange
/foo
  -> routeChange
```

Reply:
I added a fix and unit tests for all those cases